peer/brontide.go

```diff
@@ -3303,6 +3420,660 @@ func (p *Brontide) createChanCloser(channel *lnwallet.LightningChannel,
+func (p *Brontide) chanFlushEventSentinel(chanCloser *chancloser.RbfChanCloser,
+	link htlcswitch.ChannelUpdateHandler,
+	channel *lnwallet.LightningChannel) {
+
+	// If there's no link, then the channel has already been flushed, so we
+	// don't need to continue.
+	if link == nil {
+		return
+	}
+
+	coopCloseStates := chanCloser.RegisterStateEvents()
+	defer chanCloser.RemoveStateSub(coopCloseStates)
+
+	newStateChan := coopCloseStates.NewItemCreated.ChanOut()
+
+	sendChanFlushed := func() {
+		chanState := channel.StateSnapshot()
+
+		peerLog.Infof("ChannelPoint(%v) has been flushed for co-op "+
+			"close, sending event to chan closer",
+			channel.ChannelPoint())
+
+		chanBalances := chancloser.ShutdownBalances{
+			LocalBalance:  chanState.LocalBalance,
+			RemoteBalance: chanState.RemoteBalance,
+		}
+		ctx, _ := p.cg.Create(context.Background())
+		chanCloser.SendEvent(ctx, &chancloser.ChannelFlushed{
+			ShutdownBalances: chanBalances,
+			FreshFlush:       true,
+		})
+	}
+
+	// We'll wait until the channel enters the ChannelFlushing state. We
+	// exit after a success loop. As after the first RBF iteration, the
+	// channel will always be flushed.
+	for newState := range newStateChan {
+		if _, ok := newState.(*chancloser.ChannelFlushing); ok {
+			peerLog.Infof("ChannelPoint(%v): rbf coop "+
+				"close is awaiting a flushed state, "+
+				"registering with link..., ",
+				channel.ChannelPoint())
+
+			// Request the link to send the event once the channel
+			// is flushed. We only need this event sent once, so we
+			// can exit now.
+			link.OnFlushedOnce(sendChanFlushed)
+
+			return
+		}
+	}
+}
+
+// initRbfChanCloser initializes the channel closer for a channel that
+// is using the new RBF based co-op close protocol. This only creates the chan
+// closer, but doesn't attempt to trigger any manual state transitions.
+func (p *Brontide) initRbfChanCloser(
+	channel *lnwallet.LightningChannel) (*chancloser.RbfChanCloser, error) {
+
+	chanID := lnwire.NewChanIDFromOutPoint(channel.ChannelPoint())
+
+	link := p.fetchLinkFromKeyAndCid(chanID)
+
+	_, startingHeight, err := p.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain best block: %w", err)
+	}
+
+	defaultFeePerKw, err := p.cfg.FeeEstimator.EstimateFeePerKW(
+		p.cfg.CoopCloseTargetConfs,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to estimate fee: %w", err)
+	}
+
+	thawHeight, err := channel.AbsoluteThawHeight()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get thaw height: %w", err)
+	}
+
+	peerPub := *p.IdentityKey()
+
+	msgMapper := chancloser.NewRbfMsgMapper(
+		uint32(startingHeight), chanID, peerPub,
+	)
+
+	initialState := chancloser.ChannelActive{}
+
+	scid := channel.ShortChanID()
+
+	env := chancloser.Environment{
+		ChainParams:    p.cfg.Wallet.Cfg.NetParams,
+		ChanPeer:       peerPub,
+		ChanPoint:      channel.ChannelPoint(),
+		ChanID:         chanID,
```

Comment:
@Crypt-iQ this is where the `chanID` is set. What did you have in mind to prevent a caller from setting the wrong value? The env doesn't know what the "correct" value is, it relies on the caller to have an authoritative view. The caller here is peer, which is assumed to know the correct chan ID value.